src/document/printer.js

```diff
@@ -508,7 +508,7 @@ function printDocToString(doc, options) {
         // Ideally we wouldn't mutate the array here but copying all the
         // elements to a new array would make this algorithm quadratic,
         // which is unusable for large arrays (e.g. large texts in JSX).
-        parts.splice(0, 2);
+        parts = parts.toSpliced(0, 2);
```

Comment:
Ha, this is the root cause... It will be hard to fix, this function mutates array because this comment https://github.com/prettier/prettier/blob/484ecde4307d3f535256ef83e9b09c5a0445bc77/src/document/printer.js#L506

I try to solve this problem before, but didn't get time to finish https://github.com/prettier/prettier/pull/13315

Reply:
> Or is there no way to copy doc? 

Copy is not a good idea, the doc can be huge.

Let's take my approach from #13315 . It creates mutable `fill` during print.

Reply:
Sorry, I have a limited context. What work is left in that PR?

For now, I think we can add the current bug case as a test.

Reply:
Okay, I can take care of it, thanks for looking into the issue.

Reply:
Ok. I understood to some extent the goal and context that PR is trying to reach.

I think the implementation of PR is complete, is the test of performance the cause of the bottleneck? Please feel free to let me know if there is anything I can do to help!

Reply:
Yes, it's complete, but I was waiting for a better solution :smile: